File: src/libs/m3/pes/PE.cc
/*
 * Copyright (C) 2016-2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <m3/pes/PE.h>
#include <m3/session/ResMng.h>
#include <m3/Exception.h>
#include <m3/Syscalls.h>
#include <m3/pes/VPE.h>

#include <iostream>
#include <sstream>

namespace m3 {

static PEDesc desc_with_properties(PEDesc desc, const std::string &props) {
    auto res = desc;
    std::stringstream ss(props);
    std::string prop;
    while(std::getline(ss, prop, '+')) {
        if(prop == "imem")
            res = PEDesc(PEType::COMP_IMEM, res.isa(), 0);
        else if(prop == "emem" || prop == "vm")
            res = PEDesc(PEType::COMP_EMEM, res.isa(), 0);
        else if(prop == "arm")
            res = PEDesc(res.type(), PEISA::ARM, 0);
        else if(prop == "x86")
            res = PEDesc(res.type(), PEISA::X86, 0);
        else if(prop == "riscv")
            res = PEDesc(res.type(), PEISA::RISCV, 0);
        else if(prop == "rocket")
            res = PEDesc(res.type(), res.isa(), 0, res.attr() | PEAttr::ROCKET);
        else if(prop == "boom")
            res = PEDesc(res.type(), res.isa(), 0, res.attr() | PEAttr::BOOM);
        else if(prop == "nic")
            res = PEDesc(res.type(), res.isa(), 0, res.attr() | PEAttr::NIC);
        else if(prop == "indir")
            res = PEDesc(PEType::COMP_IMEM, PEISA::ACCEL_INDIR, 0);
        else if(prop == "copy")
            res = PEDesc(PEType::COMP_IMEM, PEISA::ACCEL_COPY, 0);
        else if(prop == "rot13")
            res = PEDesc(PEType::COMP_IMEM, PEISA::ACCEL_ROT13, 0);
        else if(prop == "idedev")
            res = PEDesc(PEType::COMP_IMEM, PEISA::IDE_DEV, 0);
        else if(prop == "nicdev")
            res = PEDesc(PEType::COMP_IMEM, PEISA::NIC_DEV, 0);
    }
    return res;
}

PE::~PE() {
    if(_free) {
        try {
            VPE::self().resmng()->free_pe(sel());
        }
        catch(...) {
            // ignore
        }
    }
}

Reference<PE> PE::alloc(const PEDesc &desc) {
    capsel_t sel = VPE::self().alloc_sel();
    PEDesc res = VPE::self().resmng()->alloc_pe(sel, desc);
    return Reference<PE>(new PE(sel, res, KEEP_CAP, true));
}

Reference<PE> PE::get(const std::string &desc) {
    auto own = VPE::self().pe();
    std::stringstream ss(desc);
    std::string props;
    while(std::getline(ss, props, '|')) {
        if(props == "own") {
            if(own->desc().supports_pemux() && own->desc().has_virtmem())
                return own;
        }
        else if(props == "clone") {
            try {
                return PE::alloc(own->desc());
            }
            catch(...) {
            }
        }
        else {
            auto base = PEDesc(own->desc().type(), own->desc().isa(), 0);
            try {
                return PE::alloc(desc_with_properties(base, props));
            }
            catch(...) {
            }
        }
    }
    VTHROW(Errors::NOT_FOUND, "Unable to find PE with " << desc.c_str());
}

Reference<PE> PE::derive(uint eps) {
    capsel_t sel = VPE::self().alloc_sel();
    Syscalls::derive_pe(this->sel(), sel, eps);
    return Reference<PE>(new PE(sel, desc(), 0, false));
}

uint PE::quota() const {
    return Syscalls::pe_quota(sel());
}

}

File: src/libs/rust/m3/src/pes/pe.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use core::fmt;

use crate::cap::{CapFlags, Capability, Selector};
use crate::errors::{Code, Error};
use crate::kif::PEDesc;
use crate::pes::VPE;
use crate::quota::Quota;
use crate::rc::Rc;
use crate::syscalls;
use crate::tcu::PEId;

/// Represents a processing element.
pub struct PE {
    cap: Capability,
    id: PEId,
    desc: PEDesc,
    free: bool,
}

impl PE {
    /// Allocates a new PE from the resource manager with given description
    pub fn new(desc: PEDesc) -> Result<Rc<Self>, Error> {
        let sel = VPE::cur().alloc_sel();
        let (id, ndesc) = VPE::cur().resmng().unwrap().alloc_pe(sel, desc)?;
        Ok(Rc::new(PE {
            cap: Capability::new(sel, CapFlags::KEEP_CAP),
            id,
            desc: ndesc,
            free: true,
        }))
    }

    /// Binds a new PE object to given selector
    pub fn new_bind(id: PEId, desc: PEDesc, sel: Selector) -> Self {
        PE {
            cap: Capability::new(sel, CapFlags::KEEP_CAP),
            id,
            desc,
            free: false,
        }
    }

    /// Gets a PE with given description.
    ///
    /// The description is an '|' separated list of properties that will be tried in order. Two
    /// special properties are supported:
    /// - "own" to denote the own PE (provided that it has support for multiple VPEs)
    /// - "clone" to denote a separate PE that is identical to the own PE
    ///
    /// For other properties, see `PEDesc::derive`.
    ///
    /// Examples:
    /// - PE with an arbitrary ISA, but preferred the own: "own|core"
    /// - Identical PE, but preferred a separate one: "clone|own"
    /// - BOOM core if available, otherwise any core: "boom|core"
    /// - BOOM with NIC if available, otherwise a Rocket: "boom+nic|rocket"
    pub fn get(desc: &str) -> Result<Rc<Self>, Error> {
        let own = VPE::cur().pe();
        for props in desc.split('|') {
            match props {
                "own" => {
                    if own.desc().supports_pemux() && own.desc().has_virtmem() {
                        return Ok(own.clone());
                    }
                },
                "clone" => {
                    if let Ok(pe) = Self::new(own.desc()) {
                        return Ok(pe);
                    }
                },
                p => {
                    let base = PEDesc::new(own.desc().pe_type(), own.desc().isa(), 0);
                    if let Ok(pe) = Self::new(base.with_properties(p)) {
                        return Ok(pe);
                    }
                },
            }
        }
        Err(Error::new(Code::NotFound))
    }

    /// Derives a new PE object from `self` with a subset of the resources, removing them from `self`
    ///
    /// The three resources are the number of EPs, the time slice length in nanoseconds, and the
    /// number of page tables.
    pub fn derive(
        &self,
        eps: Option<u32>,
        time: Option<u64>,
        pts: Option<u64>,
    ) -> Result<Rc<Self>, Error> {
        let sel = VPE::cur().alloc_sel();
        syscalls::derive_pe(self.sel(), sel, eps, time, pts)?;
        Ok(Rc::new(PE {
            cap: Capability::new(sel, CapFlags::empty()),
            desc: self.desc(),
            id: self.id(),
            free: false,
        }))
    }

    /// Returns the selector
    pub fn sel(&self) -> Selector {
        self.cap.sel()
    }

    /// Returns the PE id
    pub fn id(&self) -> PEId {
        self.id
    }

    /// Returns the PE description
    pub fn desc(&self) -> PEDesc {
        self.desc
    }

    /// Returns the EP, time, and page table quota
    pub fn quota(&self) -> Result<(Quota<u32>, Quota<u64>, Quota<usize>), Error> {
        syscalls::pe_quota(self.sel())
    }

    /// Sets the quota of the PE with given selector to specified initial values (given time slice
    /// length and number of page tables).
    ///
    /// This call requires a root PE capability.
    pub fn set_quota(&self, time: u64, pts: u64) -> Result<(), Error> {
        syscalls::pe_set_quota(self.sel(), time, pts)
    }
}

impl Drop for PE {
    fn drop(&mut self) {
        if self.free {
            VPE::cur().resmng().unwrap().free_pe(self.sel()).ok();
        }
    }
}

impl fmt::Debug for PE {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        write!(
            f,
            "PE{}[sel: {}, desc: {:?}]",
            self.id(),
            self.sel(),
            self.desc()
        )
    }
}
